### DISCUSSION TITLE
How do I use tables as entries?

### DISCUSSION THREAD

**Discussion by kentookura** [ibhagwan/fzf-lua#1490] [Q&A] [answered]
I have a command that returns a blob of json whicjh I am parsing into a lua table that Iwould like to pass into fzf.
The prototypical item looks something like this:

```lua
{
  route = "ex-0002.xml",                                                                                                    
  sourcePath = "/home/user/forest/trees/ex-0002.tree",                                                                     
  tags = {},                                                                                                                 
  taxon = vim.NIL,                                                                                                           
  title = "example"
}
```

I would like the entry to display both the title and the route, but open the file in vim upon selection.
If I pass in the entire table to the fzf callback, it just displays a bunch of lines like
```
table:0x......
```

If I pass in the title, it displays correctly, but how can I then use the `sourcePath` field when the item is selected?

Thanks!

Related:

https://github.com/ibhagwan/fzf-lua/issues/754

**Comment by ibhagwan** [comment, marked as answer]
See https://github.com/ibhagwan/fzf-lua/wiki, you can’t pass a table in the ways that you described (json key-value), to pass the data into `fzf_exec` you’ll need to convert the data into a list of strings (using `string.format`), I.e:
```lua
{ 
  “Line1”,
  “Line2”,
}
```

if you need access to metadata you’ll need to store it separately and have the line contain an index number (or any other identifier) so you can access the metadata, alternatively you can put the metadata into the line and hide it using `--with-nth` (read man fzf).

**Reply by kentookura** [reply]
Great, thanks!

**Reply by anandkumar89** [reply]
I table of bibliography data. I can create something like this 
```
{
"search_line2", "preview_line1"
"search_line2", "preview_line1"
"search_line2", "preview_line1"
}
```
Is it possible to attach identifier to each search line? I may need this in actions. e.g. open pdf corresponding to bib entry. 

--- 
table looks something like this. 
```
        table.insert(references, {
            display = v.alastnm .. " " .. v.year .. " " .. v.title,
            etype = v.etype,
            adate = v.accessDate or v.date or "0000-00-00 000",
            publicationTitle = v.publicationTitle,
            author = v.author,
            alastnm = v.alastnm,
            year = v.year,
            title = v.title,
            abstract = v.abstractNote,
            key = v.zotkey,
            cite = v.citekey,
        })
```
Whenever an item is selected, I'd need zotkey/citekey to perform callback functions. 

```
            attach_mappings = function(prompt_bufnr, map)
            map("i", "<C-o>", function()
                local selection = action_state.get_selected_entry()
                actions.close(prompt_bufnr)
                cb(selection.value.cite) 
            end)
            map("i", "<CR>", function()
                local selection = action_state.get_selected_entry()
                -- actions.close(prompt_bufnr)
                print(selection.value)
                require("zotcite.get").open_attachment(selection.value.cite)
            end)
            return true
            end,
```
I'm complete begginer, any help would be appreciated.

**Reply by nenahp** [reply]
If citekey is a kind of uuid? Just embed it into your line then piped to fzf. Use `--with-nth`/`--delimiter` to hide it.
```lua
local references = { ... }
local lines = vim.tbl_filter(function(v) return v.citekey .. '\t' .. v.content end, references)
require('fzf-lua').fzf_exec(lines, {
  -- don't show citekey, show content only
  fzf_opts = { ['--delimiter'] = '\t', ['--with-nth'] = '2' },
  actions = {
    enter = function(seletions, opts)
      local seletion = seletions[1]
      local citekey, content = unpack(vim.split(seletion, '|', { trimempty = true }))
      require('zotcite.get').open_attachment(citekey)
    end,
  },
})
```
If citekey is complex string you can encode it by `base64`, then decode it when do action. (Or implement a hash/lookup table by your own)

**Reply by anandkumar89** [reply]
Thank You for such a quick response :) 

citekey is simple `author2010` type string, it is unique to each line. I also want to add a preview 

<img width="1344" alt="Screenshot 2025-03-28 at 18 04 14" src="https://github.com/user-attachments/assets/a11dd77d-2249-4841-96a0-c96183e32f2e" />

Only few columns are included in `lines`, remaining details are dumped in preview as formatted string - which I can generate given hidden `citekey`. But then How can I populate preview using a function - assuming I already have `function generate_preview(citekey) ... end`.

--

**Reply by nenahp** [reply]
```lua
require('fzf-lua').fzf_exec(lines, {
  -- don't show citekey, show content only
  fzf_opts = { ['--delimiter'] = '\t', ['--with-nth'] = '2' },
  previewer = {
    _ctor = function()
      local base = require 'fzf-lua.previewer.builtin'.buffer_or_file
      local previewer = base:extend()
      function previewer:populate_preview_buf(selection)
        local citekey, content = unpack(vim.split(selection, '|', { trimempty = true }))
        local maybe_lines = generate_preview(citekey)
        local tmpbuf = self:get_tmp_buffer()
        vim.api.nvim_buf_set_lines(tmpbuf, 0, -1, false, maybe_lines)
        -- vim.bo[tmpbuf].filetype = ...
        self:set_preview_buf(tmpbuf)
      end
      return previewer
    end,
  },
  actions = { ... },
})
```

**Reply by anandkumar89** [reply]
Thanks a ton! [works](https://github.com/jalvesaq/zotcite/issues/96#issuecomment-2762257667) great 🚀

**Reply by anandkumar89** [reply]
In context of tables, how does one achieve grep_lgrep kind of behaviour via [ctrl-g]? 

i.e. I want to first search the lines for exact matches using first query and then use the results to fuzzy find. Like we do in `live_grep`. I'm trying to achieve the same but have been struggling. 

1. Do we have access to all filtered lines after query, which I can then pass to new picker for fuzzy searching without `--exact`.
2. Or we write a function that returns results for a given query, and use this for second fuzzy filtering along with `opts.last_query`  ?

or is there more elegant way?

**Reply by nenahp** [reply]
https://github.com/ibhagwan/fzf-lua/discussions/1109


Another approach is `fzf_opts['--listen'] = true`, then `vim.fn.system('curl localhost:{FZF_PORT}')` you will get a more structured response. Check detail in `man fzf | nvim +/FZF_PORT`

**Reply by anandkumar89** [reply]
Thank You 🚀
